### ISSUE TITLE
Pretrained Weights Issue and Trajectory Visualization

### ISSUE THREAD

**Issue by Vilonge** [horipse01/3d-foundation-policy#7]
Great work, and thanks for open-sourcing the code and weights! I’m currently facing two issues that need resolution:

1. Pretrained Weights Issue
When after loading the pretrained weights (include ema model) and adding LoRA layers, the loss and gradient norms show unusual magnitudes as shown in figure.
<img width="463" height="80" alt="Image" src="https://github.com/user-attachments/assets/a9d90605-919e-4bb4-8168-233feb46a352" />

2. Missing Environment for Rollout Visualization
The code includes interfaces for trajectory visualization, but the pretrained weights provide env config is None. How should I configure the environment to generate any visualizable rollout?
<img width="273" height="48" alt="Image" src="https://github.com/user-attachments/assets/03a4b391-6e7f-42f6-a357-e53555508b0c" /><img width="333" height="189" alt="Image" src="https://github.com/user-attachments/assets/ee735e7b-eff2-4358-9ed1-f458fb56b535" />

**Comment by horipse01** [comment]
For the first question, check whether you have correctly configured the accelerate library and are using deepspeed. Finally, run your script using `accelerate launch xx`. Below is the accelerate and deepspeed configuration file I use as a reference.
- accelerate config
```
compute_environment: LOCAL_MACHINE
debug: false
deepspeed_config:
  deepspeed_config_file: /cephfs/shared/yangrujia/project/droid_policy_learning/config.json
  zero3_init_flag: false
distributed_type: DEEPSPEED
downcast_bf16: 'no'
enable_cpu_affinity: false
machine_rank: 0
main_training_function: main
num_machines: 1
num_processes: 8
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```
- deepspeed config
```
{
    "fp16": {
    "enabled": false
    },
    "fp8": {
        "enabled": true,
        "compute_dtype": "bfloat16",
        "quantization_dtype": "fp8",
        "precision_mode": "fp8",
        "loss_scale": 1024
    },
    "zero_optimization": {
        "stage": 2,
        "allgather_partitions": true,
        "allgather_bucket_size": 2e8,
        "overlap_comm": true,
        "reduce_scatter": true,
        "reduce_bucket_size": 2e8,
        "contiguous_gradients": true
    },
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": 1e-4,
        "weight_decay": 0.0,
            "torch_adam": true,
            "adam_w_mode": true
        }
    },
    "scheduler": {
        "type": "WarmupDecayLR",
        "params": {
            "warmup_min_lr": 4e-5,
            "warmup_max_lr": 1e-4,
            "warmup_num_steps": 1000,
            "total_num_steps": 10000
        }
    },
    "gradient_accumulation_steps": 1,
    "gradient_clipping": "auto",
    "steps_per_print": 2000,
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": 8,
    "wall_clock_breakdown": false
}

```
For the second question, I believe the features you mentioned were implemented in robomimic mainly for simulation experiments. In our work, they are actually not used, and the rollout visualize functionality is also not available.

**Comment by Vilonge** [comment]
> For the first question, check whether you have correctly configured the accelerate library and are using deepspeed. Finally, run your script using `accelerate launch xx`. Below is the accelerate and deepspeed configuration file I use as a reference.
> 
> * accelerate config
> 
> ```
> compute_environment: LOCAL_MACHINE
> debug: false
> deepspeed_config:
>   deepspeed_config_file: /cephfs/shared/yangrujia/project/droid_policy_learning/config.json
>   zero3_init_flag: false
> distributed_type: DEEPSPEED
> downcast_bf16: 'no'
> enable_cpu_affinity: false
> machine_rank: 0
> main_training_function: main
> num_machines: 1
> num_processes: 8
> rdzv_backend: static
> same_network: true
> tpu_env: []
> tpu_use_cluster: false
> tpu_use_sudo: false
> use_cpu: false
> ```
> 
> * deepspeed config
> 
> ```
> {
>     "fp16": {
>     "enabled": false
>     },
>     "fp8": {
>         "enabled": true,
>         "compute_dtype": "bfloat16",
>         "quantization_dtype": "fp8",
>         "precision_mode": "fp8",
>         "loss_scale": 1024
>     },
>     "zero_optimization": {
>         "stage": 2,
>         "allgather_partitions": true,
>         "allgather_bucket_size": 2e8,
>         "overlap_comm": true,
>         "reduce_scatter": true,
>         "reduce_bucket_size": 2e8,
>         "contiguous_gradients": true
>     },
>     "optimizer": {
>         "type": "AdamW",
>         "params": {
>             "lr": 1e-4,
>         "weight_decay": 0.0,
>             "torch_adam": true,
>             "adam_w_mode": true
>         }
>     },
>     "scheduler": {
>         "type": "WarmupDecayLR",
>         "params": {
>             "warmup_min_lr": 4e-5,
>             "warmup_max_lr": 1e-4,
>             "warmup_num_steps": 1000,
>             "total_num_steps": 10000
>         }
>     },
>     "gradient_accumulation_steps": 1,
>     "gradient_clipping": "auto",
>     "steps_per_print": 2000,
>     "train_batch_size": "auto",
>     "train_micro_batch_size_per_gpu": 8,
>     "wall_clock_breakdown": false
> }
> ```
> 
> For the second question, I believe the features you mentioned were implemented in robomimic mainly for simulation experiments. In our work, they are actually not used, and the rollout visualize functionality is also not available.

Thanks for your reply. Regarding the first question, after implementing the configuration file you provided, the model now shows the loss values like this when test on 4 samples.

<img width="526" height="90" alt="Image" src="https://github.com/user-attachments/assets/7c539462-6a7a-4bd2-bc1a-b060b026bcc4" />

 Are these values reasonable? And could you share your final loss values in the paper?

**Comment by horipse01** [comment]
Sorry to hear it didn’t solve the issue. If you're using pretrained weights for finetuning, the loss should quickly drop below 0.1 within 100 batches, and the final loss should be less than 0.01. Please double-check that the data format and content are correct. You can also try training from scratch by not passing the `ckpt` parameter to see if the loss behaves normally.
